oxia-operator/controllers/oxiacluster_resources.go

```diff
@@ -0,0 +1,339 @@
+// Copyright 2023 StreamNative, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"fmt"
+	monitoringV1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	oxiav1alpha1 "github.com/streamnative/oxia/api/v1alpha1"
+	"golang.org/x/exp/maps"
+	"gopkg.in/yaml.v2"
+	appsV1 "k8s.io/api/apps/v1"
+	coreV1 "k8s.io/api/core/v1"
+	rbacV1 "k8s.io/api/rbac/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+)
+
+func resourceName(component Component, name string) string {
+	if component == Coordinator {
+		return name + "-coordinator"
+	}
+	return name
+}
+
+func objectMeta(component Component, oxia *oxiav1alpha1.OxiaCluster) metaV1.ObjectMeta {
+	_resourceName := resourceName(component, oxia.Name)
+	return metaV1.ObjectMeta{
+		Name:      _resourceName,
+		Namespace: oxia.Namespace,
+		Labels:    allLabels(component, oxia.Name),
+	}
+}
+
+func allLabels(component Component, name string) map[string]string {
+	_allLabels := make(map[string]string)
+	maps.Copy(_allLabels, selectorLabels(component, name))
+	//TODO inject version label
+	maps.Copy(_allLabels, additionalLabels("TODO"))
+	return _allLabels
+}
+
+func selectorLabels(component Component, name string) map[string]string {
+	_resourceName := resourceName(component, name)
+	return map[string]string{
+		"app.kubernetes.io/name":      "oxia-cluster",
+		"app.kubernetes.io/Component": string(component),
+		"app.kubernetes.io/instance":  _resourceName,
+	}
+}
+func additionalLabels(version string) map[string]string {
+	return map[string]string{
+		"app.kubernetes.io/version":    version,
+		"app.kubernetes.io/part-of":    "oxia",
+		"app.kubernetes.io/managed-by": "oxia-controller",
+	}
+}
+
+func serviceAccount(component Component, cluster *oxiav1alpha1.OxiaCluster) *coreV1.ServiceAccount {
+	_serviceAccount := &coreV1.ServiceAccount{
+		ObjectMeta: objectMeta(component, cluster),
+	}
+	if cluster.Spec.Image.PullSecrets != nil {
+		_serviceAccount.ImagePullSecrets = []coreV1.LocalObjectReference{{Name: *cluster.Spec.Image.PullSecrets}}
+	}
+	return _serviceAccount
+}
+
+func role(component Component, cluster *oxiav1alpha1.OxiaCluster) *rbacV1.Role {
```

Comment:
These RBAC roles seem to be automatically generated based on annotations
https://stackoverflow.com/questions/70664728/how-to-add-rbac-roles-to-a-controller-for-a-different-kind-of-resource-in-kubebu